spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala

```diff
@@ -401,6 +391,90 @@ class OptimizeExecutor(
     }
   }
 
+  /**
+   * Utility methods to group bins into batches for incremental progress.
+   *
+   * @param bins List of bins to group into batches
+   * @return Sequence of batches. Each batch contains one or more bins.
+   */
+  private def groupBinsIntoBatches(
+      bins: Seq[(Map[String, String], Seq[AddFile])],
+      batchSize: Long)
+  : Seq[Seq[(Map[String, String], Seq[AddFile])]] = {
+    val batches = new ArrayBuffer[Seq[(Map[String, String], Seq[AddFile])]]()
+
+    val currentBatch = new ArrayBuffer[(Map[String, String], Seq[AddFile])]()
+    var currentBatchSize = 0L
+
+    bins.foreach { bin =>
+      val binSize = bin._2.map(_.size).sum
+      currentBatch += bin
+      currentBatchSize += binSize
+
+      if (currentBatchSize > batchSize) {
+        batches += currentBatch.toVector
+        currentBatch.clear()
+        currentBatchSize = 0
+      }
+    }
+
+    if (currentBatch.nonEmpty) {
+      batches += currentBatch.toVector
+    }
+
+    batches.toSeq
+  }
+
+  private def runOptimizeBatch(
+    bins: Seq[(Map[String, String], Seq[AddFile])],
+    maxFileSize: Long
+  ): (Seq[AddFile], Seq[RemoveFile], Seq[DeletionVectorDescriptor]) = {
+    val txn = snapshot.deltaLog.startTransaction(catalogTable, Some(snapshot))
+
+    val filesToProcess = bins.flatMap(_._2)
+
+    txn.trackFilesRead(filesToProcess)
```

Comment:
This should be the only difference with the existing default behavior. Only the filtered candidates are registered with the transaction, not all matching files.